### ISSUE TITLE
Safari doesn't work with Github login.

### ISSUE THREAD

**Issue by VLegakis** [netdata/netdata#6500]
@VLegakis Safari doesn't work with Github login.

_Originally posted by @cosmix in https://github.com/netdata/netdata/issues/6459#issuecomment-512369917_

**Comment by VLegakis** [comment]
Hello @cosmix could you please describe the sign-in steps you follow for this problem including:
1) Environment you are using (link if possible)
2) Browser version
3) Error logs from your browser console

Any more information you could provide would be of great help.
Many thanks.

**Comment by cosmix** [comment]
Hello @VLegakis, thanks for opening the ticket. Here is as much info as I can provide at this time:

### Environment details

OS: macOS 10.14.5 (18F203)
Browser: Safari Version 12.1.1 (14607.2.6.1.1)

#### Additional notes: 

1. netdata is proxied through nginx.
2. The local DNS is pi-hole which filters several advertising/tracking URLs. However, on the same machine(s), Firefox and Chrome both work great, so I think this is not relevant to this issue which seems to be Safari-specific.

### Steps to reproduce:

Note: Always reproducible. These steps refer to what happens AFTER initially authorising with GitHub.

1. Visit the netdata page
2. Click on 'nodes (beta)'

### Expected result

#### After having already authorised netdata using GitHub

1. netdata.cloud console opens.

#### First time

1. netdata requests authorisation using GitHub and other methods
2. user clicks on 'authorise using GitHub'
3. netdata redirects to nodes view.

### Actual result (always on Safari)

1. New browser window opens
2. multiple redirects
3. The 'Sign in' authorisation screen appears.

### Console output:

#### On the original netdata window/tab:

`Failed to load resource: the server responded with a status of 401 ()`
https://netdata.cloud/account/sso-agent?id=1b22d80a-f138-11e6-aea7-f23c91935f9f

`Unhandled Promise Rejection: TypeError: null is not an object (evaluating 'registryAgents.concat')` syncAgents — main.js:4994

**Comment by Azein** [comment]
I reproduced the issue partly, I don't get redirected, but auth is failing for me and I'm stuck in console application with empty screen.

Safari Version 10.1.1 (12603.2.4)

I believe that the root cause ( some credentials call failing) is the same to what @cosmix is experiencing.

**Comment by cosmix** [comment]
Since I'm also using 12.1.1 @VLegakis, it seems that the browser version is irrelevant here. i.e. the problem manifests itself on both 12.1.1 and 10.1.1.

**Comment by cakrit** [comment]
Please also see https://github.com/netdata/netdata/issues/6380#issuecomment-508889551
It has the same exception.

**Comment by Azein** [comment]
@cosmix 

The root cause of the issue for me is the same - failed sso agent call,

`Failed to load resource: the server responded with a status of 401 ()`

We will look into the logs.

**Comment by cosmix** [comment]
Hello @Azein — agreed. But why does the SSO fail on my and your cases and works on @VLegakis 's?

**Comment by Azein** [comment]
I was able to resolve the issue with cookies not being sent to SSO on my older Safari, also verified it on 12.1.1 We will deploy this small fix for cross-domain cookies, expecting that at least
`Failed to load resource: the server responded with a status of 401` problem will be solved.

**Comment by VLegakis** [comment]
Well, I'm not sure this is related to ReCaptcha. 
I tried it with emptying cache, relaunching, logging out and anything and the cookies get stored correctly for me => I get logged in correctly. 
Indeed we have no persuading explanation on why is it working on my Safari
I will merge the PR and see.

**Comment by Azein** [comment]
Issue isn't reproducible for me on Safari 12.x anymore, and the cookies attachment fix is delayed due to complications on staging environment. We will look into ReCaptcha thing. 

For now, I see at least two possible sources of the problem:

1) Cookies are not being set (we need a list of cookie keys from the client experiencing the issue)
2) Cookies are not being attached. This could be due to cross-domain cookies inclusion, or some other issue, like time precision difference in user's local env.

tl/dr: Fix is postponed, further investigation needed / cookie keys list from client could be useful.

**Comment by VLegakis** [comment]
Hello @cosmix,
We have deployed a possible fix for your issue. We are not 100% this will fix it since we were not able to reproduce it, but please take a a look and let us know with these new deployment.
Many thanks.

**Comment by cakrit** [comment]
Guys, try with a brand new GitHub account, in case the issue has to do with sign-up, instead of sign-in. See also https://github.com/netdata/netdata/issues/6459#issuecomment-514248263

**Comment by VLegakis** [comment]
Hello @cakrit I have tested this with a new Github account on Safari and (apart from the redirection issue #6542 which is being handled) its working fine.

**Comment by cakrit** [comment]
Yes, we had trouble reproducing my issue on Firefox as well, after the first two times when it happened. It looked like a race condition, @nicolasparada and @syedomair would look into the code, to see if they could find something suspicious. In the cases where it failed, I saw `sign-out` being called, without pressing the sign-out button.

**Comment by nicolasparada** [comment]
We looked into the code and couldn't find anything wrong. For me, the issue as reproducible last week but with the latest changes (deployed on Monday) it's fine now. Nor Netdata agents or Cloud code does a call to sign-out as side-effect, it only happens from a user action (click).

I cannot test Safari since I don't have a mac, but @Azein was also able to reproduce and its fix solved it for him.

I was thinking that maybe Google App Engine could still be serving old code for certain users (some sort of cache). But the fixes where deployed Monday, they should be already up to date...

For the case of netdata agents running through reverse-proxies, that's something that will require more work for sure. We are working on that.

**Comment by nicolasparada** [comment]
Since our change was for javascript it's very much possible that the old code is still cached.

**Comment by gpapamathND** [comment]
@cosmix could you please provide feedback? Is it still reproducible?

**Comment by cosmix** [comment]
JS caching shouldn't be an issue in any modern app. Most bundlers deal with it, as does react itself. I'll try to reproduce and report my findings here. Thanks!

**Comment by cosmix** [comment]
After testing this, I've found that:

1. When netdata is proxied via nginx
2. Netdata is accessible on port {80,443} at a URL path other than root (e.g. http://example.net/netdata)

Although Chrome and Firefox work well, Safari does not. The resulting behaviour, however, has changed however, since originally reporting this issue: it currently redirects to the root instead of the proper URL.

**Comment by cakrit** [comment]
@jacekkolasa this is one of the issues created, related to Safari. I'll close the other one and refer to this one.
